### PR TITLE
DSC-464:466/menu-constrained-inversion-sampling

### PR DESCRIPTION
**Pull request by marccjansen**
This works currently! Only some slow-down of model fit as was to be expected with an extra loop through menus. 

However, a couple points for discussion:
* I'm now questioning whether we have to just have one scope of all 'active' recipes over the training data horizon instead of a scope per menu. The reason why is that for a particular menu scope the negative item ids might be correct, but the positive item ids are still sampled from all positive interactions over the horizon, so it might still result in invalid inversions.
* A second thing to consider is whether we should shuffle the order of the menus over which we do the scoped inversion sampling. Of course this is only relevant if we don't increase the scope to all active recipes over the training horizon.

### CODE REVIEW

File: lightfm/_lightfm_fast.pyx.template [hunk 1]
```diff
@@ -857,9 +861,9 @@ def fit_warp(CSRMatrix item_features,
             while sampled < lightfm.max_sampled:
 
                 sampled = sampled + 1
-                negative_item_id = (rand_r(&random_states[{thread_num}])
-                                    % item_features.rows)
-                print("Negative item id sampled")
+                negative_item_ix = (rand_r(&random_states[{thread_num}])
+                                    % len(menu_recipe_ids))
+                negative_item_id = item_ids[negative_item_ix]
```
Comment (marccjansen) [inline lightfm/_lightfm_fast.pyx.template]: This is incorrect as is. Item_ids contain duplicates (i.e. for nnz column entry there is a row value, as per coo format), so I need to pass an array of the unique item_ids that coincide with a menu and set intersect that with item_ids and take the id at random index from the resulting set.

Comment (marccjansen) [reply lightfm/_lightfm_fast.pyx.template]: The correct menu_item_id is passed down into compute_representation, where the get_row_start and get_row_end methods are called to find the range of relevant CSR indices

File: lightfm/_lightfm_fast.pyx.template [hunk 2]
```diff
@@ -800,14 +800,15 @@ def fit_warp(CSRMatrix item_features,
...
-    cdef int negative_item_id, sampled, row
+    cdef int negative_item_ix, negative_item_id, sampled, row
     cdef double positive_prediction, negative_prediction
     cdef double loss, MAX_LOSS
     cdef flt weight
     cdef flt *user_repr
     cdef flt *pos_it_repr
     cdef flt *neg_it_repr
     cdef unsigned int[::1] random_states
+    cdef unsigned int[::1] menu_recipe_ids
```
Comment (marccjansen) [inline lightfm/_lightfm_fast.pyx.template]: Instead of typing the variable here, for the next ticket I move this into the method arguments where it is passed down from within the _run_epoch method in lightfm.py

File: lightfm/lightfm.py [hunk 3]
```diff
@@ -601,20 +606,26 @@ def _run_epoch(self, item_features, user_features, interactions,
...
-                     lightfm_data,
-                     self.learning_rate,
-                     self.item_alpha,
-                     self.user_alpha,
-                     num_threads,
-                     self.random_state)
+
+            # Cycle over scopes
+            for i in range(scope.shape[0]):
+                scope_item_ids = scope[0].indices
```
Comment (marccjansen) [inline lightfm/lightfm.py]: I'm questioning whether we have to just have one scope of all 'active' recipes over the training data horizon instead of a scope per menu. The reason why is that for a particular menu scope the negative item ids might be correct, but the positive item ids are still sampled from all positive interactions over the horizon, so it might still result in invalid inversions.

**Comment by depet** [comment]
Interesting. Let's try to discuss it the first thing tomorrow morning before my flight.

**Comment by depet** [comment]
@marc: should `c` files be in git? Aren't they generated from `pyx` files?

@depet: I followed the suggestion on Cython documentation (http://cython.readthedocs.io/en/latest/src/reference/compilation.html)  "It is strongly recommended that you distribute the generated .c files as well as your Cython sources, so that users can install your module without needing to have Cython available. It is also recommended that Cython compilation not be enabled by default in the version you distribute. Even if the user has Cython installed, he/she probably doesn’t want to use it just to install your module. Also, the installed version may not be the same one you used, and may not compile your sources correctly."

File: lightfm/lightfm.py [hunk 4]
```diff
@@ -418,7 +418,7 @@ def _check_input_finite(self, data):
             raise ValueError('Not all input values are finite. '
                              'Check the input for NaNs and infinite values.')
 
-    def fit(self, interactions,
+    def fit(self, interactions, scope=None, match_indices=None,
```
Comment (depet) [inline lightfm/lightfm.py:421]: Can you please update docstring? No `scope` and `match_indices` - I guess those args replaced `menu`?

**Review by HikariNoMJ14 — APPROVED** [review]

**Comment by depet** [comment]
Okay, interesting. I thought that c files depends on operating system etc., but if that's what they recommend I guess it is fine :)

File: lightfm/lightfm.py [hunk 5]
```diff
@@ -601,20 +617,29 @@ def _run_epoch(self, item_features, user_features, interactions,
...
-                     shuffle_indices,
-                     lightfm_data,
-                     self.learning_rate,
-                     self.item_alpha,
-                     self.user_alpha,
-                     num_threads,
-                     self.random_state)
+
+            # Cycle over scopes
+            # TODO catch scope=None and match_indices=None
```
Comment (depet) [inline lightfm/lightfm.py]: What happens if `scope is None` and `match_indices is NOT None`?

Comment (marccjansen) [reply lightfm/lightfm.py]: good question :) I would say scope is None will set match_indices to None even if match_indices are passed, because if scope is None we should revert to the original sampling in lightfm

**Review by depet — APPROVED** [review]
